### PR TITLE
differentiating between `--x` and `--x: bool`

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description
Fixes: #10450 

This pr differentiating between `--x: bool` and `--x`

Here are examples which demostrate difference between them:
```nushell
def a [--x: bool] { $x };
a --x    # not allowed, you need to parse a value to the flag.
a        # it's allowed, and the value of `$x` is false, which behaves the same to `def a [--x] { $x }; a`
```

For boolean flag with default value, it works a little bit different to #10450 mentioned:
```nushell
def foo [--option: bool = false] { $option }
foo                  # output false
foo --option         # not allowed, you need to parse a value to the flag.
foo --option true    # output true
```

# User-Facing Changes
After the pr, the following code is not allowed:
```nushell
def a [--x: bool] { $x }; a --x
```

Instead, you have to pass a value to flag `--x` like `a --x false`.  But bare flag works in the same way as before.  

## Update: one more breaking change to help on #7260 
```
def foo [--option: bool] { $option == null }
foo
```
After the pr, if we don't use a boolean flag, the value will be `null` instead of `true`.  Because here `--option: bool` is treated as a flag rather than a switch

### CODE REVIEW

File: crates/nu-parser/src/parser.rs
Comment (vector-of-bool) [inline crates/nu-parser/src/parser.rs:3718]: Should `bool` still be special-cased here? The original special casing is what led to the odd behavior discussed in #7260 ([my comment](https://github.com/nushell/nushell/issues/7260#issuecomment-1728337485)) and complicated the logic that eventually led to #10410 and then #10450.

This still leads to a one-off special behavior that

```nu
def foo [--x: T] { $x | describe }
foo
```

prints `nothing` in all cases _except_ for `T = bool`, and leaves no way to leave a boolean argument as un-specified by the caller (i.e. `foo` cannot discern between `(foo)` and `(foo --x=false)`).

**Comment by amtoine** [comment]
ooh i see what this is doing :open_mouth: 
thanks for the quick response @WindSoilder :pray: 

i'm not sure i am a big fan of requiring to pass `true` or `false` to `-x: bool` or `-x = false`, but i think i get the idea and i could buy that :yum: 

as soon as something has a type, either explicitely with `x: foo` or inferred with `x = foo`, it needs a value after, even though it's a boolean one :open_mouth: 

boolean flags, a.k.a. switches, a.k.a. flags, are then special and can only be defined with `--option`, without any default value :thinking: 

yeah, i can buy that :+1: 

:exclamation: however this is a breaking change, all the options / switches that were declared with `: bool` or ` = false` to give them a default value, will break :scream:

**Comment by fdncred** [comment]
Just adding this here because it helped me understand the changes.
```nushell
❯ def foo [-a, --long, --both(-b), -c: bool, -d: bool = true, --eee(-e): bool = true] {
  echo $"-a = ($a)"
  echo $"--long = ($long)"
  echo $"--both(-b) = ($both)"
  echo $"-c = ($c)"
  echo $"-d = ($d)"
  echo $"--eee = ($eee)"
} 
```
This PR now requires that `-c`, `-d`, and `--eee/-e` now have `true` or `false` passed in on the command line if you want to change the default. e.g. `foo -d true`

File: crates/nu-parser/src/parser.rs
Comment (WindSoilder) [reply crates/nu-parser/src/parser.rs:3718]: Hi, thanks for your comment!

Yeah I'm also thinking about what's the value of `$x` in `foo` call.  But finally I decide to make it false because it doesn't introduce a larger breaking change.  But yeah, your opinion is valid, currently I'm not really sure about it..

Comment (WindSoilder) [reply crates/nu-parser/src/parser.rs]: After a discussion with core team, I change the value to `null` to make it consistent with other types.

**Comment by fdncred** [comment]
Doesn't it seem like having this `-c: bool` default to null is going to be confusing?
Running my script above, I get this
```
> foo
-a = false
--long = false
--both-b = false
-c =
-d = true
--eee = true
```

I'm trying to figure out the rules. This is what I have so far.
* `-a` will default to false
* `--long` will default to false
* `--both(-b)` will default to false
* `-c: bool` will default to null 
* `-d: bool = true` will default to specified default value
* `--eee(-e): bool = true` will default to specified default value

Correct me if I'm wrong, maybe the reasoning that `-c: bool` defaults to null is because `-f: int` defaults to null?

**Comment by WindSoilder** [comment]
> maybe the reasoning that -c: bool defaults to null is because -f: int defaults to null?

Yeah, you're right, so in this way, we have two type of flag, one is switch(which works the same), one is something like keyword argument(once it have explicit type like `-c: bool`)

**Comment by amtoine** [comment]
i think `-b: bool` or `-s: string` defaulting to `null` is ok :yum:

**Review by amtoine — APPROVED** [review]
@WindSoilder 
i went and fixed the `toolkit.nu` because it was broken :+1: 

i was also able to fix all my scripts with this new PR and quite like the change :ok_hand: 

i personally aprove this :ok_hand:

**Comment by amtoine** [comment]
ooh i hit a wall try to fix another one of my scripts, what does that mean @WindSoilder? :confused: 

```nushell
> def foo [-x: bool] { if not $x { print "ok" }}; foo
Error: nu::shell::type_mismatch

  × Type mismatch.
   ╭─[entry #3:1:1]
 1 │ def foo [-x: bool] { if not $x { print "ok" }}; foo
   ·                             ─┬
   ·                              ╰── bool
   ╰────
```

**Comment by jntrnr** [comment]
@WindSoilder - I think for your example:

```
def foo [--bar: bool] { $bar }
```

If you call:
```
foo
```

Then `$bar` should be `null` as it is with other flags with arguments.

**Comment by amtoine** [comment]
@jntrnr 
it is with this PR :yum:
```nushell
> def foo [--bar: bool] { $bar | describe }; foo
nothing
```

**Review by fdncred — APPROVED** [review]
+1 from me to land this PR. Thanks @WindSoilder!!!

**Comment by amtoine** [comment]
@WindSoilder 
it's up to you if you wanna look at my bug above or address it later when i can investigate more after landing :relieved: :wink:

**Comment by fdncred** [comment]
@amtoine I couldn't use `if not $x` either and had to change it to something like this
```nushell
❯ def foo [-x: bool] { if ($x != true) { print "ok" }}
```
You would think that since `$x` is a bool that we'd be able to say `not $x`. I wonder why we cannot?

**Comment by WindSoilder** [comment]
> @WindSoilder it's up to you if you wanna look at my bug above or address it later when i can investigate more after landing 😌 😉

@amtoine It's because when you ran `cargo run -- ~/documents/repos/github.com/goatfiles/scripts/nu_scripts/scripts/tmux-sessionizer.nu list-sessions`, you ran `main list-sessions` function without `-m` flag, so the value of `$more` is null, and then `list-sessions` function doesn't handle `null` value.
When you use a flag(`--more(-m): boll`), the value of `$more` can be `null`, we need to handle for it

**Comment by amtoine** [comment]
ooh wow, appreciate you investigating the actual script @WindSoilder :pray: 

this one is actually quite subtle :open_mouth: 
the general rule of thumb is **never define `--switch: bool` in the public API** :+1: 

let's go ahead and land this, i'll file an issue with the unhelpful error just after i upgrade :wink:

**Comment by amtoine** [comment]
because the changes can be a bit tricky when not used to them, i think we need to mention this in the release notes of 0.86.0 :relieved:

**Pull request merged**
